osbuild/meta.py

```diff
@@ -0,0 +1,416 @@
+
+        If the schema information itself is missing, it
+        will return a `ValidationResult` in failed state,
+        with 'missing schema information' as the reason.
+        """
+        res = ValidationResult(self.name)
+
+        if not self.data:
+            return res.fail("missing schema information")
+
+        for error in self._validator.iter_errors(target):
+            res += ValidationError.from_exception(error)
+
+        return res
+
+    def __bool__(self):
+        return self.data
+
+
+class StageInfo:
+    """Meta information about a stage
+
+    Represents the information about a osbuild pipeline
+    stage or assembler (here also considered to be a stage).
+    Contains the short description (`desc`), a longer
+    description (`info`) and the JSON schema of valid options
+    (`opts`). The `validate` method will check a the options
+    of a stage instance against the JSON schema.
+
+    Normally this class is instantiated via its `load` method.
+    """
+
+    def __init__(self, klass: str, name: str, info: str):
+        self.name = name
+        self.type = klass
+
+        opts = info.get('STAGE_OPTS') or ""
+        self.info = info.get("STAGE_INFO")
+        self.desc = info.get('STAGE_DESC')
+        self.opts = info.get('STAGE_OPTS')
+        self.opts = json.loads("{" + opts + "}")
+
+    @property
+    def schema(self):
+        schema = {
+            "title": f"Pipeline {self.type}",
+            "type": "object",
+            "additionalProperties": False,
+            "properties": {
+                "name": {"type": "string"},
+                "options": {
+                    "type": "object",
+                    **self.opts
+                }
+            },
+            "required": ["name"]
+        }
+        return schema
+
+    @classmethod
+    def load(cls, root, klass, name) -> Optional["StageInfo"]:
+        names = ['STAGE_INFO', 'STAGE_DESC', 'STAGE_OPTS']
+
+        def value(a):
+            v = a.value
+            if isinstance(v, ast.Str):
+                return v.s
+            return ""
+
+        def filter_type(lst, target):
+            return [x for x in lst if isinstance(x, target)]
+
+        def targets(a):
+            return [t.id for t in filter_type(a.targets, ast.Name)]
+
+        mapping = {
+            "Stage": "stages",
+            "Assembler": "assemblers"
+        }
+
+        if klass not in mapping:
+            raise ValueError(f"Unsupported type: {klass}")
+
+        base = mapping[klass]
+
+        path = os.path.join(root, base, name)
+        try:
+            with open(path) as f:
+                data = f.read()
+        except FileNotFoundError:
+            return None
+
+        tree = ast.parse(data, name)
+        assigns = filter_type(tree.body, ast.Assign)
+        targets = [(t, a) for a in assigns for t in targets(a)]
+        info = {k: value(v) for k, v in targets if k in names}
+        return cls(klass, name, info)
```

Comment:
So we now require the stages to be python scripts? If that is the case, why do all this dance, instead of just turning stages into proper python modules with well-known classes or functions that return us this information?

This is more of a general question. I am fine keeping it as it is to move forward.

Reply:
This is just a pragmatic approach. I have a (stale) local branch with code that actually adds a different `main` to all stages/sources/assembler that handles `--info`, `--schema` and returns the respective information. Once that lands, adapting that function to call the "binary" and obtain the information should not be too hard.

Reply:
General comment: Ideally we don't tie ourselves to python too much. By that I mean, if we don't assume python at all: great. If we do assume python but in a way that wouldn't be too hard to work around in the future when someone decides it is time for a rust/bash stage, I'd be perfectly fine with that too if that makes things simpler in the short turm. But I'd hesitate to tie it tightly to python by making things proper modules etc. At least that's my gut feeling.